x/axelarnet/message_handler.go

```diff
@@ -0,0 +1,285 @@
+		}
+		return nil
+	default:
+		return fmt.Errorf("unrecognized message type")
+	}
+}
+
+// OnRecvMessage handles general message from a cosmos chain
+func OnRecvMessage(ctx sdk.Context, k keeper.Keeper, ibcK keeper.IBCKeeper, n types.Nexus, b types.BankKeeper, packet ibcexported.PacketI) ibcexported.Acknowledgement {
+	ack := channeltypes.NewResultAcknowledgement([]byte{byte(1)})
+
+	data, err := toICS20Packet(packet)
+	if err != nil {
+		return channeltypes.NewErrorAcknowledgement(err)
+	}
+
+	// skip if packet not sent to Axelar message sender account
+	if data.GetReceiver() != types.MessageSender.String() {
+		return ack
+	}
+
+	var msg Message
+	if err := json.Unmarshal([]byte(data.GetMemo()), &msg); err != nil {
+		ackErr := sdkerrors.Wrapf(types.ErrGeneralMessage, "cannot unmarshal memo")
+		return channeltypes.NewErrorAcknowledgement(ackErr)
+	}
+
+	// extract token from packet
+	token := extractTokenFromPacketData(packet)
+
+	path := types.NewIBCPath(packet.GetDestPort(), packet.GetDestChannel())
+
+	if err := validateMessage(ctx, ibcK, n, path, msg, token); err != nil {
+		return channeltypes.NewErrorAcknowledgement(err)
+	}
+
+	sourceChain := funcs.MustOk(k.GetChainNameByIBCPath(ctx, path))
+	sourceAddress := nexus.CrossChainAddress{
+		Chain:   funcs.MustOk(n.GetChain(ctx, sourceChain)),
+		Address: data.GetSender(),
+	}
+
+	switch msg.Type {
+	case nexus.TypeGeneralMessage:
+		err = handleMessage(ctx, n, sourceAddress, msg)
+	case nexus.TypeGeneralMessageWithToken:
+		err = handleMessageWithToken(ctx, n, b, ibcK, sourceAddress, msg, token)
+	case nexus.TypeSendToken:
+		err = handleTokenSent(ctx, n, b, ibcK, sourceAddress, msg, token)
+	default:
+		err = sdkerrors.Wrapf(types.ErrGeneralMessage, "unrecognized Message type")
+	}
+
+	if err != nil {
+		return channeltypes.NewErrorAcknowledgement(err)
+	}
+
+	return ack
+}
+
+func handleMessage(ctx sdk.Context, n types.Nexus, sourceAddress nexus.CrossChainAddress, msg Message) error {
+	destChain := funcs.MustOk(n.GetChain(ctx, nexus.ChainName(msg.DestinationChain)))
+
+	recipient := nexus.CrossChainAddress{Chain: destChain, Address: msg.DestinationAddress}
+	m := nexus.NewGeneralMessage(
+		n.GenerateMessageID(ctx, ctx.TxBytes()),
+		sourceAddress,
+		recipient,
+		crypto.Keccak256Hash(msg.Payload).Bytes(),
+		nexus.Sent,
+		nil,
+	)
+
+	events.Emit(ctx, &types.ContractCallSubmitted{
+		MessageID:        m.ID,
+		Sender:           m.Sender.Address,
+		SourceChain:      m.Sender.Chain.Name,
+		DestinationChain: m.Recipient.Chain.Name,
+		ContractAddress:  m.Recipient.Address,
+		PayloadHash:      m.PayloadHash,
+		Payload:          msg.Payload,
+	})
+
+	return n.SetNewMessage(ctx, m)
+}
+
+func handleMessageWithToken(ctx sdk.Context, n types.Nexus, b types.BankKeeper, ibcK keeper.IBCKeeper, sourceAddress nexus.CrossChainAddress, msg Message, token sdk.Coin) error {
+	destChain := funcs.MustOk(n.GetChain(ctx, nexus.ChainName(msg.DestinationChain)))
+
```

Comment:
restrict destination to be EVM for call contract/call contract with token until we require ExecuteMessage to be called for routing